File: babbleon/manifest.py
"""
Tool manifest: the list of real names that Babbleon tracks per host.

Community edition: static default list, or a local TOML file.
Enterprise extension: MDM-pushed manifest, fleet-policy inventory,
per-team override — loaded via the babbleon.manifest_sources entry_point.

Manifest format (TOML):
  [manifest]
  version = 1
  tracked = ["curl", "ssh", ...]
  # optional: per-tool metadata (enterprise only)
  [tools.curl]
  tier = "must"
  credential_env = []
"""

from __future__ import annotations

import pathlib

try:
    import tomllib
except ImportError:
    import tomli as tomllib  # type: ignore[no-redef]  # Python <3.11 fallback


DEFAULT_TRACKED: list[str] = [
    "curl", "ssh", "nc", "python3", "bash", "wget", "git",
    "aws", "gh", "kubectl", "docker", "terraform", "npm", "pip",
]


class Manifest:
    def __init__(self, tracked: list[str]) -> None:
        self.tracked = tracked

    @classmethod
    def default(cls) -> "Manifest":
        return cls(tracked=list(DEFAULT_TRACKED))

    @classmethod
    def from_file(cls, path: pathlib.Path) -> "Manifest":
        data = tomllib.loads(path.read_text())
        tracked = data.get("manifest", {}).get("tracked", DEFAULT_TRACKED)
        return cls(tracked=tracked)

    @classmethod
    def load(cls, path: pathlib.Path | None = None) -> "Manifest":
        """Load from file if present, otherwise use defaults."""
        if path and path.exists():
            return cls.from_file(path)
        return cls.default()

File: crates/babbleon/src/manifest.rs
//! Tracked-tool manifest.
//!
//! Community edition: static default or local TOML. Enterprise edition:
//! MDM-pushed manifest via the plugin registry. Same `Manifest` type either way.

use crate::Result;
use serde::Deserialize;
use std::path::Path;

pub const DEFAULT_TRACKED: &[&str] = &[
    "curl", "ssh", "nc", "python3", "bash", "wget", "git",
    "aws", "gh", "kubectl", "docker", "terraform", "npm", "pip",
];

#[derive(Debug, Clone)]
pub struct Manifest {
    pub tracked: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct ManifestFile {
    manifest: ManifestSection,
}

#[derive(Debug, Deserialize)]
struct ManifestSection {
    #[serde(default)]
    tracked: Vec<String>,
}

impl Manifest {
    pub fn default_tracked() -> Self {
        Self {
            tracked: DEFAULT_TRACKED.iter().map(|s| s.to_string()).collect(),
        }
    }

    pub fn from_file(path: &Path) -> Result<Self> {
        let text = std::fs::read_to_string(path)?;
        let f: ManifestFile = toml::from_str(&text)?;
        let tracked = if f.manifest.tracked.is_empty() {
            DEFAULT_TRACKED.iter().map(|s| s.to_string()).collect()
        } else {
            f.manifest.tracked
        };
        Ok(Self { tracked })
    }

    /// Load from file if present, otherwise return defaults.
    pub fn load(path: Option<&Path>) -> Result<Self> {
        match path {
            Some(p) if p.exists() => Self::from_file(p),
            _ => Ok(Self::default_tracked()),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defaults_nonempty() {
        let m = Manifest::default_tracked();
        assert!(!m.tracked.is_empty());
    }

    #[test]
    fn load_falls_back_to_default() {
        let m = Manifest::load(Some(Path::new("/nonexistent.toml"))).unwrap();
        assert_eq!(m.tracked.len(), DEFAULT_TRACKED.len());
    }

    #[test]
    fn from_file_parses_toml() {
        let dir = tempfile::tempdir().unwrap();
        let p = dir.path().join("m.toml");
        std::fs::write(
            &p,
            "[manifest]\nversion = 1\ntracked = [\"foo\", \"bar\"]\n",
        )
        .unwrap();
        let m = Manifest::from_file(&p).unwrap();
        assert_eq!(m.tracked, vec!["foo", "bar"]);
    }
}
